sklearn/feature_extraction/text.py

```diff
@@ -44,6 +45,70 @@
            'strip_tags']
 
 
+def _preprocess(doc, accent_function=None, lower=False):
+    """Chain together an optional series of text preprocessing steps to
+    apply to a document.
+
+    Parameters
+    ----------
+    doc: str
+        The string to preprocess
+    accent_function: callable
+        Function for handling accented characters. Common strategies include
+        normalizing and removing.
+    lower: bool
+        Whether to use str.lower to lowercase all fo the text
+
+    Returns
+    -------
+    doc: str
+        preprocessed string
+    """
+    if lower:
+        doc = doc.lower()
+    if accent_function is not None:
+        doc = accent_function(doc)
+    return doc
+
+
+def _analyze(doc, analyzer=None, tokenizer=None, ngrams=None,
+             preprocessor=None, decoder=None, stop_words=None):
+    """Chain together an optional series of text processing steps to go from
+    a single document to ngrams, with or without tokenizing or preprocessing.
+
+    If analyzer is used, it happens at the very end, and in this codebase is
+    intended to replace the preprocessor, tokenizer, and ngrams step.
+
+    Parameters
+    ----------
+    analyzer: callable
+    tokenizer: callable
+    ngrams: callable
+    preprocessor: callable
+    decoder: callable
+    stop_words: list
+
+    Returns
+    -------
+    ngrams: list
+        A sequence of tokens, possibly with pairs, triples, etc.
+    """
+    if decoder is not None:
+        doc = decoder(doc)
+    if preprocessor is not None:
+        doc = preprocessor(doc)
+    if tokenizer is not None:
+        doc = tokenizer(doc)
+    if ngrams is not None:
+        if stop_words is not None:
+            doc = ngrams(doc, stop_words)
+        else:
+            doc = ngrams(doc)
+    if analyzer is not None:
+        doc = analyzer(doc)
```

Comment:
If an analyzer is provided, non of the other functions should  be run.

I'm not very keen on adding such function that obfuscates the code flow logic in order to emulate function piping. Though I agree with you with the fact that there may not be a better way.

Reply:
Yeah 😔.

I've made the analyzer if/else with preprocess+tokenize+ngrams.